File: sdk/python/src/budgetary/_internal/http.py
"""Single chokepoint for HTTP. Handles auth, error mapping, and retry orchestration."""

from __future__ import annotations

import json
import math
import time
from email.utils import parsedate_to_datetime
from typing import Any, Callable, TypeVar

import httpx

from budgetary.errors import (
    BudgetaryAuthError,
    BudgetaryError,
    BudgetaryNetworkError,
    BudgetaryNotFoundError,
    BudgetaryPermissionError,
    BudgetaryRateLimitError,
    BudgetaryServerError,
    BudgetaryValidationError,
)

from .retry import with_retry

T = TypeVar("T")

# Hard ceiling on a response body (8 MiB). The API's real responses are a few KB;
# a much larger body from a hostile or misbehaving endpoint is a memory-exhaustion
# vector, so the body is streamed and bounded rather than buffered whole.
MAX_RESPONSE_BYTES = 8 * 1024 * 1024


def _reject_non_finite(token: str) -> Any:
    """`json.loads` parse_constant hook: refuse the bare ``Infinity`` /
    ``-Infinity`` / ``NaN`` tokens from a hostile server instead of letting them
    flow into the numeric dataclasses."""
    raise ValueError(f"non-finite JSON constant: {token}")


def _reject_non_finite_float(token: str) -> float:
    """`json.loads` parse_float hook: refuse a numeric literal that OVERFLOWS to
    infinity (e.g. ``1e400``) — ``parse_constant`` only sees the bare keyword
    tokens, so this closes the overflow-to-``inf`` path into the dataclasses."""
    value = float(token)
    if not math.isfinite(value):
        raise ValueError(f"non-finite JSON number: {token}")
    return value


def _default_code_for_status(status: int) -> str:
    return {
        400: "invalid_request",
        401: "authentication_failed",
        403: "permission_denied",
        404: "not_found",
        409: "idempotency_conflict",
        413: "payload_too_large",
        429: "rate_limited",
        503: "unavailable",
    }.get(status, "internal_error" if status >= 500 else f"http_{status}")


def _parse_retry_after(header: str | None) -> float | None:
    if header is None:
        return None
    try:
        return float(header)
    except (TypeError, ValueError):
        pass
    try:
        target = parsedate_to_datetime(header)
    except (TypeError, ValueError):
        return None
    if target is None:
        return None
    import time

    seconds = target.timestamp() - time.time()
    return max(0.0, seconds)


def _build_error(
    status: int, body: dict[str, Any] | None, headers: httpx.Headers
) -> BudgetaryError:
    err_block = (body or {}).get("error") or {}
    code = err_block.get("code") or _default_code_for_status(status)
    message = err_block.get("message") or f"Budgetary API returned HTTP {status}"
    request_id = err_block.get("request_id") or headers.get("x-request-id")

    kwargs = dict(code=code, message=message, http_status=status, request_id=request_id)

    if status == 401:
        return BudgetaryAuthError(**kwargs)
    if status == 403:
        # Distinct from 401 (bad key): a valid key lacking scope. A sibling of
        # BudgetaryAuthError (both under BudgetaryError), so the two are
        # distinguishable — matching the TS SDK. See contract §6.
        return BudgetaryPermissionError(**kwargs)
    if status == 404:
        return BudgetaryNotFoundError(**kwargs)
    if status in (400, 409, 413):
        return BudgetaryValidationError(**kwargs)
    if status == 429:
        return BudgetaryRateLimitError(
            **kwargs,
            retry_after_seconds=_parse_retry_after(headers.get("retry-after")),
        )
    if status >= 500:
        return BudgetaryServerError(**kwargs)
    return BudgetaryError(**kwargs)


class HttpClient:
    def __init__(
        self,
        *,
        api_key: str,
        base_url: str,
        timeout_ms: int,
        max_retries: int,
        http_client: httpx.Client | None = None,
    ) -> None:
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._timeout_ms = timeout_ms
        self._max_retries = max_retries
        self._owned_client = http_client is None
        self._client = http_client or httpx.Client()

    def close(self) -> None:
        if self._owned_client:
            self._client.close()

    def request(
        self,
        *,
        method: str,
        path: str,
        json_body: Any | None = None,
        params: dict[str, Any] | None = None,
        timeout_ms: int | None = None,
        parse: Callable[[Any], T],
    ) -> T:
        return with_retry(
            lambda: self._attempt(
                method=method,
                path=path,
                json_body=json_body,
                params=params,
                timeout_ms=timeout_ms,
                parse=parse,
            ),
            max_retries=self._max_retries,
        )

    def _attempt(
        self,
        *,
        method: str,
        path: str,
        json_body: Any | None,
        params: dict[str, Any] | None,
        timeout_ms: int | None,
        parse: Callable[[Any], T],
    ) -> T:
        url = self._base_url + path
        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "Accept": "application/json",
        }
        if json_body is not None:
            headers["Content-Type"] = "application/json"

        timeout_seconds = (timeout_ms or self._timeout_ms) / 1000.0

        clean_params: dict[str, Any] | None = None
        if params is not None:
            clean_params = {}
            for key, value in params.items():
                if value is None:
                    continue
                if isinstance(value, bool):
                    clean_params[key] = "true" if value else "false"
                else:
                    clean_params[key] = value

        # httpx's timeout is per-operation (connect / read / write), so a slow
        # drip could still run past it in wall-clock terms. Bound the whole
        # request with a monotonic deadline enforced while streaming the body.
        deadline = time.monotonic() + timeout_seconds
        raw = bytearray()
        try:
            with self._client.stream(
                method,
                url,
                headers=headers,
                json=json_body,
                params=clean_params,
                timeout=timeout_seconds,
                # Never follow a redirect: a hostile 3xx would re-send the request
                # (and the Authorization header) to its `Location`. Set explicitly
                # so a caller-supplied client with `follow_redirects=True` can't
                # re-enable it.
                follow_redirects=False,
            ) as response:
                declared = response.headers.get("content-length")
                if declared is not None:
                    try:
                        if int(declared) > MAX_RESPONSE_BYTES:
                            raise BudgetaryNetworkError(
                                code="network",
                                message="response body from Budgetary API exceeds the size limit",
                            )
                    except ValueError:
                        pass  # unparseable Content-Length: fall back to the streamed cap
                for chunk in response.iter_bytes():
                    raw.extend(chunk)
                    if len(raw) > MAX_RESPONSE_BYTES:
                        raise BudgetaryNetworkError(
                            code="network",
                            message="response body from Budgetary API exceeds the size limit",
                        )
                    if time.monotonic() > deadline:
                        raise BudgetaryNetworkError(
                            code="timeout",
                            message="request to Budgetary API exceeded its total deadline",
                        )
                status_code = response.status_code
                resp_headers = response.headers
        except httpx.TimeoutException as err:
            raise BudgetaryNetworkError(
                code="timeout", message=f"request to Budgetary API timed out: {err}"
            ) from err
        except httpx.ConnectError as err:
            raise BudgetaryNetworkError(
                code="network",
                message=f"could not connect to Budgetary API: {err}",
            ) from err
        except httpx.RequestError as err:
            raise BudgetaryNetworkError(
                code="network",
                message=f"network error while contacting Budgetary API: {err}",
            ) from err

        body: dict[str, Any] | None = None
        if raw:
            try:
                # `parse_constant` rejects Infinity/NaN; `errors="replace"` turns a
                # non-UTF-8 body into text that then fails to parse (→ body=None),
                # matching the prior "non-JSON body" handling.
                body = json.loads(
                    raw.decode("utf-8", errors="replace"),
                    parse_constant=_reject_non_finite,
                    parse_float=_reject_non_finite_float,
                )
            except ValueError:
                body = None

        if status_code >= 400:
            raise _build_error(status_code, body, resp_headers)

        try:
            return parse(body or {})
        except (TypeError, KeyError, AttributeError) as err:
            # A 2xx whose body is missing, non-JSON, a non-object JSON value
            # (list/str/number → AttributeError on `.get`), or missing a required
            # field is a truncated or malformed response — surface it inside the
            # error taxonomy as a network-class failure rather than letting a raw
            # exception escape. Additive/unknown fields are tolerated by the
            # parsers, so this only fires on a genuinely unusable body.
            raise BudgetaryNetworkError(
                code="network",
                message=f"unusable response body from Budgetary API: {err!r}",
            ) from err

File: sdk/typescript/src/internal/http.ts
import {
  BudgetaryAuthError,
  BudgetaryError,
  BudgetaryNetworkError,
  BudgetaryNotFoundError,
  BudgetaryPermissionError,
  BudgetaryRateLimitError,
  BudgetaryServerError,
  BudgetaryValidationError,
} from "../errors.js";
import { withRetry, type RetryInfo } from "./retry.js";

export interface HttpClientConfig {
  apiKey: string;
  baseUrl: string;
  timeoutMs: number;
  maxRetries: number;
  fetchImpl?: typeof fetch;
  /** Observe each retry (additive; forwarded to the retry wrapper). */
  onRetry?: (info: RetryInfo) => void;
}

export interface HttpRequest {
  method: "GET" | "POST";
  path: string;
  body?: unknown;
  query?: Record<string, string | number | boolean | undefined | null>;
  timeoutMs?: number;
  /**
   * Optional caller cancellation, combined with the per-attempt timeout so a
   * host-abandoned request aborts the in-flight fetch AND stops the retry ladder.
   */
  signal?: AbortSignal;
}

const NETWORK_ERROR_MESSAGE = "network error while contacting Budgetary API";
const TIMEOUT_ERROR_MESSAGE = "request to Budgetary API timed out";

/**
 * Hard ceiling on a response body (8 MiB). The API's real responses are a few KB;
 * a much larger body from a hostile or misbehaving endpoint is a memory-exhaustion
 * vector, so the read is bounded and a body over the cap is rejected as a network
 * error rather than buffered in full.
 */
const MAX_RESPONSE_BYTES = 8 * 1024 * 1024;

/**
 * Read a response body with a hard byte cap. Rejects up-front on an oversized
 * `Content-Length`, and otherwise streams and aborts once the cap is exceeded so
 * a lying/absent `Content-Length` cannot force an unbounded buffer. Falls back to
 * `response.text()` when the environment/test double exposes no readable stream.
 */
async function readBodyCapped(response: Response): Promise<string> {
  const declared = response.headers.get("content-length");
  if (declared !== null) {
    const n = Number(declared);
    if (Number.isFinite(n) && n > MAX_RESPONSE_BYTES) {
      throw new BudgetaryNetworkError({
        code: "network",
        message: "response body from Budgetary API exceeds the size limit",
      });
    }
  }
  const oversize = () =>
    new BudgetaryNetworkError({
      code: "network",
      message: "response body from Budgetary API exceeds the size limit",
    });

  const body = response.body;
  if (!body || typeof body.getReader !== "function") {
    // No readable stream (e.g. a test double): read fully, then enforce by size.
    const text = await response.text();
    if (new TextEncoder().encode(text).length > MAX_RESPONSE_BYTES) throw oversize();
    return text;
  }
  const reader = body.getReader();
  const chunks: Uint8Array[] = [];
  let total = 0;
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    if (value) {
      total += value.byteLength;
      if (total > MAX_RESPONSE_BYTES) {
        await reader.cancel().catch(() => {});
        throw oversize();
      }
      chunks.push(value);
    }
  }
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return new TextDecoder("utf-8").decode(merged);
}

function snakeKey(key: string): string {
  return key.replace(/([A-Z])/g, (m) => `_${m.toLowerCase()}`);
}

function camelKey(key: string): string {
  return key.replace(/_([a-z0-9])/g, (_, c: string) => c.toUpperCase());
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (value === null || typeof value !== "object") return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

/** A container the key-transform walks recurse into: an array or a plain object. */
function isContainer(
  value: unknown,
): value is unknown[] | Record<string, unknown> {
  return Array.isArray(value) || isPlainObject(value);
}

/**
 * Remove trailing slashes from a base URL with a single linear scan.
 *
 * The obvious `value.replace(/\/+$/, "")` is quadratic: the start of `\/+` is
 * unanchored, so on input like `"…" + "/".repeat(n) + "x"` the engine retries
 * the greedy slash run from each of the n slash positions (O(n^2)) — the
 * "polynomial regular expression" ReDoS class. This loop is O(n) and carries
 * no backtracking, regardless of how the base URL was supplied.
 */
function stripTrailingSlashes(value: string): string {
  let end = value.length;
  while (end > 0 && value.charCodeAt(end - 1) === 0x2f /* "/" */) end--;
  return value.slice(0, end);
}

export function toSnakeCase(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(toSnakeCase);
  }
  if (isPlainObject(value)) {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value)) {
      // `metadata` is a caller-owned, free-form map (contract §4.2): its keys are
      // data, not protocol, so they must reach the wire byte-for-byte — never
      // snake-cased. Every other key is a known protocol field and is recursed.
      out[snakeKey(k)] = k === "metadata" ? v : toSnakeCase(v);
    }
    return out;
  }
  return value;
}

/**
 * Deep-camelCase a parsed response body, ITERATIVELY. A recursive walk throws a
 * raw `RangeError` (stack overflow) on a deeply-nested body from a hostile or
 * misbehaving endpoint — escaping the SDK's error taxonomy. An explicit worklist
 * keeps the traversal on the heap, so depth is bounded only by the already-capped
 * 8 MiB body, never by the call stack. Behavior is otherwise identical: array
 * and plain-object keys are camelCased; every other value passes through.
 */
export function toCamelCase(value: unknown): unknown {
  if (!isContainer(value)) return value;
  const root: unknown[] | Record<string, unknown> = Array.isArray(value)
    ? []
    : {};
  const stack: Array<
    [unknown[] | Record<string, unknown>, unknown[] | Record<string, unknown>]
  > = [[value, root]];
  while (stack.length > 0) {
    const [src, dst] = stack.pop()!;
    if (Array.isArray(src)) {
      const out = dst as unknown[];
      for (let i = 0; i < src.length; i++) {
        const v = src[i];
        if (isContainer(v)) {
          const child: unknown[] | Record<string, unknown> = Array.isArray(v)
            ? []
            : {};
          out[i] = child;
          stack.push([v, child]);
        } else {
          out[i] = v;
        }
      }
    } else {
      const out = dst as Record<string, unknown>;
      for (const [k, v] of Object.entries(src as Record<string, unknown>)) {
        const nk = camelKey(k);
        if (isContainer(v)) {
          const child: unknown[] | Record<string, unknown> = Array.isArray(v)
            ? []
            : {};
          out[nk] = child;
          stack.push([v, child]);
        } else {
          out[nk] = v;
        }
      }
    }
  }
  return root;
}

interface WireErrorBody {
  error?: {
    code?: string;
    message?: string;
    request_id?: string;
  };
}

function parseRetryAfter(header: string | null): number | null {
  if (header === null) return null;
  const n = Number(header);
  if (Number.isFinite(n)) return n;
  // HTTP date format fallback: header may be an HTTP-date.
  const t = Date.parse(header);
  if (!Number.isNaN(t)) {
    const seconds = Math.max(0, Math.round((t - Date.now()) / 1000));
    return seconds;
  }
  return null;
}

function buildError(
  status: number,
  body: WireErrorBody | null,
  headers: Headers,
): BudgetaryError {
  const code = body?.error?.code ?? defaultCodeForStatus(status);
  const message =
    body?.error?.message ?? `Budgetary API returned HTTP ${status}`;
  const requestId =
    body?.error?.request_id ?? headers.get("x-request-id") ?? null;
  const args = { code, message, httpStatus: status, requestId };

  if (status === 401) return new BudgetaryAuthError(args);
  if (status === 403) return new BudgetaryPermissionError(args);
  if (status === 404) return new BudgetaryNotFoundError(args);
  if (status === 400 || status === 409 || status === 413) {
    return new BudgetaryValidationError(args);
  }
  if (status === 429) {
    return new BudgetaryRateLimitError({
      ...args,
      retryAfterSeconds: parseRetryAfter(headers.get("retry-after")),
    });
  }
  if (status >= 500) return new BudgetaryServerError(args);
  return new BudgetaryError(args);
}

function defaultCodeForStatus(status: number): string {
  switch (status) {
    case 400:
      return "invalid_request";
    case 401:
      return "authentication_failed";
    case 403:
      return "permission_denied";
    case 404:
      return "not_found";
    case 409:
      return "idempotency_conflict";
    case 413:
      return "payload_too_large";
    case 429:
      return "rate_limited";
    case 503:
      return "unavailable";
    default:
      return status >= 500 ? "internal_error" : `http_${status}`;
  }
}

function buildQueryString(
  query: HttpRequest["query"] | undefined,
): string {
  if (!query) return "";
  const params = new URLSearchParams();
  for (const [k, v] of Object.entries(query)) {
    if (v === undefined || v === null) continue;
    params.append(snakeKey(k), String(v));
  }
  const s = params.toString();
  return s.length > 0 ? `?${s}` : "";
}

export class HttpClient {
  private readonly config: HttpClientConfig;
  private readonly fetchImpl: typeof fetch;

  constructor(config: HttpClientConfig) {
    this.config = config;
    this.fetchImpl = config.fetchImpl ?? fetch;
  }

  async request<T>(req: HttpRequest): Promise<T> {
    const result = await withRetry(() => this.attempt(req), {
      maxRetries: this.config.maxRetries,
      onRetry: this.config.onRetry,
      // Stop the retry ladder (and cut a backoff sleep short) when the caller aborts.
      signal: req.signal,
    });
    return result as T;
  }

  private async attempt(req: HttpRequest): Promise<unknown> {
    const url =
      stripTrailingSlashes(this.config.baseUrl) +
      req.path +
      buildQueryString(req.query);
    // The target host (never the path/query — those can carry ids) for a
    // diagnosable transport error message.
    const host = safeHost(this.config.baseUrl);

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.config.apiKey}`,
      Accept: "application/json",
    };

    let body: string | undefined;
    if (req.body !== undefined) {
      body = JSON.stringify(toSnakeCase(req.body));
      headers["Content-Type"] = "application/json";
    }

    const timeoutMs = req.timeoutMs ?? this.config.timeoutMs;
    // Combine the per-attempt timeout with the caller's cancellation: whichever
    // fires first aborts the fetch. `AbortSignal.any` propagates the winning
    // signal's reason, so a timeout still surfaces as a TimeoutError (→ "timeout")
    // and a host abort as an AbortError (→ "abort") in mapNetworkError.
    const timeoutSignal = AbortSignal.timeout(timeoutMs);
    const signal =
      req.signal !== undefined
        ? AbortSignal.any([timeoutSignal, req.signal])
        : timeoutSignal;

    let response: Response;
    try {
      response = await this.fetchImpl(url, {
        method: req.method,
        headers,
        body,
        signal,
        // A hostile endpoint could answer with a 3xx to a `Location` it controls;
        // following it would re-POST the request body (and the Authorization
        // header) to that host. Refuse redirects outright (parity with the Python
        // SDK's httpx default of `follow_redirects=False`).
        redirect: "error",
      });
    } catch (err) {
      throw mapNetworkError(err, host);
    }

    let text: string;
    try {
      text = await readBodyCapped(response);
    } catch (err) {
      if (err instanceof BudgetaryError) throw err;
      // The body read can stall or time out after the headers arrived (the
      // AbortSignal also covers the body). Classify it as a network error —
      // exactly like a failure of the fetch call itself — instead of letting a
      // raw undici/stream error escape unclassified.
      throw mapNetworkError(err, host);
    }

    let parsed: unknown = null;
    if (text.length > 0) {
      try {
        parsed = JSON.parse(text);
      } catch {
        if (response.ok) {
          throw new BudgetaryNetworkError({
            code: "network",
            message: "received non-JSON response from Budgetary API",
          });
        }
        parsed = null;
      }
    }

    if (!response.ok) {
      throw buildError(
        response.status,
        parsed as WireErrorBody | null,
        response.headers,
      );
    }

    // `JSON.parse` rejects the bare `Infinity`/`NaN` tokens but silently coerces
    // an overflowing literal (`1e400`) to `Infinity`. Refuse any non-finite
    // number so it can't reach the numeric response fields.
    assertFiniteNumbers(parsed);
    return toCamelCase(parsed);
  }
}

/**
 * Throw a network error if any number anywhere in a (successful) response body is
 * non-finite. ITERATIVE (explicit worklist) so a deeply-nested body can't blow
 * the call stack with a raw `RangeError`. Applied only to 2xx bodies — an error
 * body's numbers are never read as numbers.
 */
function assertFiniteNumbers(root: unknown): void {
  const stack: unknown[] = [root];
  while (stack.length > 0) {
    const value = stack.pop();
    if (typeof value === "number") {
      if (!Number.isFinite(value)) {
        throw new BudgetaryNetworkError({
          code: "network",
          message: "response from Budgetary API contained a non-finite number",
        });
      }
    } else if (Array.isArray(value)) {
      for (const v of value) stack.push(v);
    } else if (value !== null && typeof value === "object") {
      for (const v of Object.values(value as Record<string, unknown>)) {
        stack.push(v);
      }
    }
  }
}

/** The target host (`host[:port]`) for an error message, or `undefined` if unparseable. */
function safeHost(baseUrl: string): string | undefined {
  try {
    return new URL(baseUrl).host || undefined;
  } catch {
    return undefined;
  }
}

/** Append ` (host: X)` when a host is known. Host only — never the path/query. */
function withHost(base: string, host?: string): string {
  return host ? `${base} (host: ${host})` : base;
}

function mapNetworkError(err: unknown, host?: string): BudgetaryError {
  if (err instanceof Error) {
    if (err.name === "TimeoutError") {
      return new BudgetaryNetworkError({
        code: "timeout",
        message: withHost(TIMEOUT_ERROR_MESSAGE, host),
      });
    }
    if (err.name === "AbortError") {
      return new BudgetaryNetworkError({
        code: "abort",
        message: "request was aborted",
      });
    }
    // Undici wraps the REAL reason (ECONNREFUSED / ENOTFOUND / a TLS failure) in
    // `err.cause`, while `err.message` is often just the opaque "fetch failed".
    // Surface the cause message + the target host so a transport failure is
    // diagnosable rather than a dead end. (Python already interpolates this; the
    // TS path previously dropped both.)
    const cause =
      err.cause instanceof Error && err.cause.message
        ? ` (${err.cause.message})`
        : "";
    return new BudgetaryNetworkError({
      code: "network",
      message: `${withHost(NETWORK_ERROR_MESSAGE, host)}: ${err.message}${cause}`,
    });
  }
  return new BudgetaryNetworkError({
    code: "network",
    message: withHost(NETWORK_ERROR_MESSAGE, host),
  });
}
